File: packages/core/src/state-trie.ts
import { hash as cryptoHash } from './crypto.js';

export interface TrieNode {
  hash: string;
  key?: string;
  value?: unknown;
  children: Map<string, TrieNode>;
}

export interface MerkleProofPath {
  key: string;
  value: unknown;
  proof: string[];
  index: number;
}

export class StateTrie {
  private root: TrieNode;
  private storage: Map<string, unknown> = new Map();

  constructor() {
    this.root = { hash: '', children: new Map() };
  }

  async set(contractId: string, key: string, value: unknown): Promise<void> {
    const fullKey = `${contractId}:${key}`;
    this.storage.set(fullKey, value);
    await this.updateRoot();
  }

  async get(contractId: string, key: string): Promise<unknown> {
    const fullKey = `${contractId}:${key}`;
    return this.storage.get(fullKey);
  }

  async delete(contractId: string, key: string): Promise<void> {
    const fullKey = `${contractId}:${key}`;
    this.storage.delete(fullKey);
    await this.updateRoot();
  }

  async getContractState(contractId: string): Promise<Record<string, unknown>> {
    const prefix = `${contractId}:`;
    const state: Record<string, unknown> = {};
    
    for (const [key, value] of this.storage) {
      if (key.startsWith(prefix)) {
        const localKey = key.slice(prefix.length);
        state[localKey] = value;
      }
    }
    
    return state;
  }

  async setContractState(contractId: string, state: Record<string, unknown>): Promise<void> {
    const prefix = `${contractId}:`;
    for (const key of this.storage.keys()) {
      if (key.startsWith(prefix)) {
        this.storage.delete(key);
      }
    }
    
    for (const [key, value] of Object.entries(state)) {
      this.storage.set(`${prefix}${key}`, value);
    }
    
    await this.updateRoot();
  }

  private async updateRoot(): Promise<void> {
    const sortedKeys = Array.from(this.storage.keys()).sort();
    const leaves: string[] = [];
    
    for (const key of sortedKeys) {
      const value = this.storage.get(key);
      const leafData = JSON.stringify({ key, value });
      const leafHash = await cryptoHash(leafData);
      leaves.push(leafHash);
    }
    
    if (leaves.length === 0) {
      this.root.hash = await cryptoHash('empty');
      return;
    }
    
    this.root.hash = await this.computeMerkleRoot(leaves);
  }

  private async computeMerkleRoot(leaves: string[]): Promise<string> {
    if (leaves.length === 0) return cryptoHash('empty');
    if (leaves.length === 1) return leaves[0];
    
    const nextLevel: string[] = [];
    for (let i = 0; i < leaves.length; i += 2) {
      const left = leaves[i];
      const right = leaves[i + 1] || left;
      const combined = await cryptoHash(left + right);
      nextLevel.push(combined);
    }
    
    return this.computeMerkleRoot(nextLevel);
  }

  async getRoot(): Promise<string> {
    return this.root.hash || await cryptoHash('empty');
  }

  async getProof(contractId: string, key: string): Promise<MerkleProofPath | null> {
    const fullKey = `${contractId}:${key}`;
    const value = this.storage.get(fullKey);
    
    if (value === undefined) return null;
    
    const sortedKeys = Array.from(this.storage.keys()).sort();
    const index = sortedKeys.indexOf(fullKey);
    
    if (index === -1) return null;
    
    const leaves: string[] = [];
    for (const k of sortedKeys) {
      const v = this.storage.get(k);
      const leafData = JSON.stringify({ key: k, value: v });
      const leafHash = await cryptoHash(leafData);
      leaves.push(leafHash);
    }
    
    const proof = await this.generateProof(leaves, index);
    
    return {
      key: fullKey,
      value,
      proof,
      index
    };
  }

  private async generateProof(leaves: string[], targetIndex: number): Promise<string[]> {
    const proof: string[] = [];
    let currentLevel = leaves;
    let idx = targetIndex;
    
    while (currentLevel.length > 1) {
      const siblingIdx = idx % 2 === 0 ? idx + 1 : idx - 1;
      
      if (siblingIdx < currentLevel.length) {
        proof.push(currentLevel[siblingIdx]);
      } else {
        proof.push(currentLevel[idx]);
      }
      
      const nextLevel: string[] = [];
      for (let i = 0; i < currentLevel.length; i += 2) {
        const left = currentLevel[i];
        const right = currentLevel[i + 1] || left;
        const combined = await cryptoHash(left + right);
        nextLevel.push(combined);
      }
      
      currentLevel = nextLevel;
      idx = Math.floor(idx / 2);
    }
    
    return proof;
  }

  async verifyProof(
    key: string,
    value: unknown,
    proof: string[],
    index: number,
    expectedRoot: string
  ): Promise<boolean> {
    const leafData = JSON.stringify({ key, value });
    let currentHash = await cryptoHash(leafData);
    let idx = index;
    
    for (const sibling of proof) {
      if (idx % 2 === 0) {
        currentHash = await cryptoHash(currentHash + sibling);
      } else {
        currentHash = await cryptoHash(sibling + currentHash);
      }
      idx = Math.floor(idx / 2);
    }
    
    return currentHash === expectedRoot;
  }

  async computeEffectsHash(changes: { key: string; preValue: unknown; postValue: unknown }[]): Promise<string> {
    const sorted = changes.sort((a, b) => a.key.localeCompare(b.key));
    const data = JSON.stringify(sorted);
    return cryptoHash(data);
  }

  toJSON(): { storage: [string, unknown][] } {
    return {
      storage: Array.from(this.storage.entries())
    };
  }

  static async fromJSON(data: { storage: [string, unknown][] }): Promise<StateTrie> {
    const trie = new StateTrie();
    for (const [key, value] of data.storage) {
      trie.storage.set(key, value);
    }
    await trie.updateRoot();
    return trie;
  }

  async clone(): Promise<StateTrie> {
    const newTrie = new StateTrie();
    for (const [key, value] of this.storage) {
      newTrie.storage.set(key, JSON.parse(JSON.stringify(value)));
    }
    await newTrie.updateRoot();
    return newTrie;
  }
}

export class ReceiptsTrie {
  private receipts: Map<string, unknown> = new Map();
  private root: string = '';

  async addReceipt(callId: string, receipt: unknown): Promise<void> {
    this.receipts.set(callId, receipt);
    await this.updateRoot();
  }

  async getReceipt(callId: string): Promise<unknown> {
    return this.receipts.get(callId);
  }

  private async updateRoot(): Promise<void> {
    const sortedIds = Array.from(this.receipts.keys()).sort();
    const leaves: string[] = [];
    
    for (const id of sortedIds) {
      const receipt = this.receipts.get(id);
      const leafData = JSON.stringify({ id, receipt });
      const leafHash = await cryptoHash(leafData);
      leaves.push(leafHash);
    }
    
    if (leaves.length === 0) {
      this.root = await cryptoHash('empty-receipts');
      return;
    }
    
    this.root = await this.computeMerkleRoot(leaves);
  }

  private async computeMerkleRoot(leaves: string[]): Promise<string> {
    if (leaves.length === 0) return cryptoHash('empty-receipts');
    if (leaves.length === 1) return leaves[0];
    
    const nextLevel: string[] = [];
    for (let i = 0; i < leaves.length; i += 2) {
      const left = leaves[i];
      const right = leaves[i + 1] || left;
      const combined = await cryptoHash(left + right);
      nextLevel.push(combined);
    }
    
    return this.computeMerkleRoot(nextLevel);
  }

  async getRoot(): Promise<string> {
    if (!this.root) {
      this.root = await cryptoHash('empty-receipts');
    }
    return this.root;
  }

  async getProof(callId: string): Promise<MerkleProofPath | null> {
    const receipt = this.receipts.get(callId);
    if (!receipt) return null;
    
    const sortedIds = Array.from(this.receipts.keys()).sort();
    const index = sortedIds.indexOf(callId);
    
    if (index === -1) return null;
    
    const leaves: string[] = [];
    for (const id of sortedIds) {
      const r = this.receipts.get(id);
      const leafData = JSON.stringify({ id, receipt: r });
      const leafHash = await cryptoHash(leafData);
      leaves.push(leafHash);
    }
    
    const proof = await this.generateProof(leaves, index);
    
    return {
      key: callId,
      value: receipt,
      proof,
      index
    };
  }

  private async generateProof(leaves: string[], targetIndex: number): Promise<string[]> {
    const proof: string[] = [];
    let currentLevel = leaves;
    let idx = targetIndex;
    
    while (currentLevel.length > 1) {
      const siblingIdx = idx % 2 === 0 ? idx + 1 : idx - 1;
      
      if (siblingIdx < currentLevel.length) {
        proof.push(currentLevel[siblingIdx]);
      } else {
        proof.push(currentLevel[idx]);
      }
      
      const nextLevel: string[] = [];
      for (let i = 0; i < currentLevel.length; i += 2) {
        const left = currentLevel[i];
        const right = currentLevel[i + 1] || left;
        const combined = await cryptoHash(left + right);
        nextLevel.push(combined);
      }
      
      currentLevel = nextLevel;
      idx = Math.floor(idx / 2);
    }
    
    return proof;
  }

  clear(): void {
    this.receipts.clear();
    this.root = '';
  }

  toJSON(): { receipts: [string, unknown][] } {
    return {
      receipts: Array.from(this.receipts.entries())
    };
  }

  static async fromJSON(data: { receipts: [string, unknown][] }): Promise<ReceiptsTrie> {
    const trie = new ReceiptsTrie();
    for (const [key, value] of data.receipts) {
      trie.receipts.set(key, value);
    }
    await trie.updateRoot();
    return trie;
  }
}

File: packages/rinku-node/src/state_trie.rs
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use serde_json::Value;
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MerkleProofPath {
    pub key: String,
    pub value: Value,
    pub proof: Vec<String>,
    pub index: usize,
}

#[derive(Debug, Clone)]
pub struct StateTrie {
    storage: HashMap<String, Value>,
    root_hash: String,
}

fn compute_hash(data: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data.as_bytes());
    hex::encode(hasher.finalize())
}

impl StateTrie {
    pub fn new() -> Self {
        StateTrie {
            storage: HashMap::new(),
            root_hash: compute_hash("empty"),
        }
    }

    pub fn set(&mut self, contract_id: &str, key: &str, value: Value) {
        let full_key = format!("{}:{}", contract_id, key);
        self.storage.insert(full_key, value);
        self.update_root();
    }

    pub fn get(&self, contract_id: &str, key: &str) -> Option<&Value> {
        let full_key = format!("{}:{}", contract_id, key);
        self.storage.get(&full_key)
    }

    pub fn delete(&mut self, contract_id: &str, key: &str) {
        let full_key = format!("{}:{}", contract_id, key);
        self.storage.remove(&full_key);
        self.update_root();
    }

    pub fn get_contract_state(&self, contract_id: &str) -> HashMap<String, Value> {
        let prefix = format!("{}:", contract_id);
        let mut state = HashMap::new();

        for (key, value) in &self.storage {
            if key.starts_with(&prefix) {
                let local_key = &key[prefix.len()..];
                state.insert(local_key.to_string(), value.clone());
            }
        }

        state
    }

    pub fn set_contract_state(&mut self, contract_id: &str, state: &HashMap<String, Value>) {
        let prefix = format!("{}:", contract_id);

        let keys_to_remove: Vec<String> = self
            .storage
            .keys()
            .filter(|k| k.starts_with(&prefix))
            .cloned()
            .collect();

        for key in keys_to_remove {
            self.storage.remove(&key);
        }

        for (key, value) in state {
            self.storage
                .insert(format!("{}{}", prefix, key), value.clone());
        }

        self.update_root();
    }

    fn update_root(&mut self) {
        let mut sorted_keys: Vec<&String> = self.storage.keys().collect();
        sorted_keys.sort();

        if sorted_keys.is_empty() {
            self.root_hash = compute_hash("empty");
            return;
        }

        let leaves: Vec<String> = sorted_keys
            .iter()
            .map(|key| {
                let value = self.storage.get(*key).unwrap();
                compute_hash(&format!("{}:{}", key, value))
            })
            .collect();

        self.root_hash = compute_merkle_root(&leaves);
    }

    pub fn root_hash(&self) -> &str {
        &self.root_hash
    }

    pub fn get_proof(&self, contract_id: &str, key: &str) -> Option<MerkleProofPath> {
        let full_key = format!("{}:{}", contract_id, key);
        let value = self.storage.get(&full_key)?;

        let mut sorted_keys: Vec<&String> = self.storage.keys().collect();
        sorted_keys.sort();

        let index = sorted_keys.iter().position(|k| **k == full_key)?;

        let leaves: Vec<String> = sorted_keys
            .iter()
            .map(|k| {
                let v = self.storage.get(*k).unwrap();
                compute_hash(&format!("{}:{}", k, v))
            })
            .collect();

        let proof = compute_merkle_proof(&leaves, index);

        Some(MerkleProofPath {
            key: key.to_string(),
            value: value.clone(),
            proof,
            index,
        })
    }

    pub fn verify_proof(&self, proof: &MerkleProofPath, contract_id: &str) -> bool {
        let full_key = format!("{}:{}", contract_id, proof.key);
        let leaf_hash = compute_hash(&format!("{}:{}", full_key, proof.value));

        verify_merkle_proof(&leaf_hash, &proof.proof, proof.index, &self.root_hash)
    }

    pub fn size(&self) -> usize {
        self.storage.len()
    }

    pub fn to_json(&self) -> StateTrieSnapshot {
        StateTrieSnapshot {
            storage: self.storage.clone(),
            root_hash: self.root_hash.clone(),
        }
    }

    pub fn from_json(snapshot: StateTrieSnapshot) -> Self {
        StateTrie {
            storage: snapshot.storage,
            root_hash: snapshot.root_hash,
        }
    }
}

impl Default for StateTrie {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StateTrieSnapshot {
    pub storage: HashMap<String, Value>,
    pub root_hash: String,
}

fn compute_merkle_root(leaves: &[String]) -> String {
    if leaves.is_empty() {
        return compute_hash("empty");
    }

    if leaves.len() == 1 {
        return leaves[0].clone();
    }

    let mut current_layer = leaves.to_vec();

    while current_layer.len() > 1 {
        let mut next_layer = Vec::new();

        let mut i = 0;
        while i < current_layer.len() {
            let left = &current_layer[i];
            let right = if i + 1 < current_layer.len() {
                &current_layer[i + 1]
            } else {
                left
            };

            next_layer.push(compute_hash(&format!("{}{}", left, right)));
            i += 2;
        }

        current_layer = next_layer;
    }

    current_layer[0].clone()
}

fn compute_merkle_proof(leaves: &[String], index: usize) -> Vec<String> {
    if leaves.len() <= 1 {
        return vec![];
    }

    let mut proof = Vec::new();
    let mut current_layer = leaves.to_vec();
    let mut idx = index;

    while current_layer.len() > 1 {
        let sibling_idx = if idx % 2 == 0 { idx + 1 } else { idx - 1 };

        if sibling_idx < current_layer.len() {
            proof.push(current_layer[sibling_idx].clone());
        } else {
            proof.push(current_layer[idx].clone());
        }

        let mut next_layer = Vec::new();
        let mut i = 0;
        while i < current_layer.len() {
            let left = &current_layer[i];
            let right = if i + 1 < current_layer.len() {
                &current_layer[i + 1]
            } else {
                left
            };

            next_layer.push(compute_hash(&format!("{}{}", left, right)));
            i += 2;
        }

        idx /= 2;
        current_layer = next_layer;
    }

    proof
}

fn verify_merkle_proof(leaf_hash: &str, proof: &[String], index: usize, expected_root: &str) -> bool {
    let mut current = leaf_hash.to_string();
    let mut idx = index;

    for sibling in proof {
        let (left, right) = if idx % 2 == 0 {
            (&current, sibling)
        } else {
            (sibling, &current)
        };

        current = compute_hash(&format!("{}{}", left, right));
        idx /= 2;
    }

    current == expected_root
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_set_and_get() {
        let mut trie = StateTrie::new();

        trie.set("contract1", "balance", json!(100));
        trie.set("contract1", "owner", json!("alice"));

        assert_eq!(trie.get("contract1", "balance"), Some(&json!(100)));
        assert_eq!(trie.get("contract1", "owner"), Some(&json!("alice")));
        assert_eq!(trie.get("contract1", "unknown"), None);
    }

    #[test]
    fn test_delete() {
        let mut trie = StateTrie::new();

        trie.set("contract1", "key", json!("value"));
        assert!(trie.get("contract1", "key").is_some());

        trie.delete("contract1", "key");
        assert!(trie.get("contract1", "key").is_none());
    }

    #[test]
    fn test_contract_state() {
        let mut trie = StateTrie::new();

        trie.set("contract1", "a", json!(1));
        trie.set("contract1", "b", json!(2));
        trie.set("contract2", "c", json!(3));

        let state = trie.get_contract_state("contract1");
        assert_eq!(state.len(), 2);
        assert_eq!(state.get("a"), Some(&json!(1)));
        assert_eq!(state.get("b"), Some(&json!(2)));
    }

    #[test]
    fn test_root_hash_changes() {
        let mut trie = StateTrie::new();
        let empty_root = trie.root_hash().to_string();

        trie.set("contract1", "key", json!("value"));
        let root1 = trie.root_hash().to_string();

        trie.set("contract1", "key2", json!("value2"));
        let root2 = trie.root_hash().to_string();

        assert_ne!(empty_root, root1);
        assert_ne!(root1, root2);
    }

    #[test]
    fn test_proof_generation_and_verification() {
        let mut trie = StateTrie::new();

        trie.set("contract1", "a", json!(1));
        trie.set("contract1", "b", json!(2));
        trie.set("contract1", "c", json!(3));

        let proof = trie.get_proof("contract1", "b").unwrap();

        assert!(trie.verify_proof(&proof, "contract1"));
        assert_eq!(proof.key, "b");
        assert_eq!(proof.value, json!(2));
    }

    #[test]
    fn test_snapshot() {
        let mut trie = StateTrie::new();
        trie.set("contract1", "key", json!("value"));

        let snapshot = trie.to_json();
        let restored = StateTrie::from_json(snapshot);

        assert_eq!(restored.root_hash(), trie.root_hash());
        assert_eq!(
            restored.get("contract1", "key"),
            Some(&json!("value"))
        );
    }
}
